lib/events/complete.go

```diff
@@ -167,3 +174,86 @@ func (u *UploadCompleter) Close() error {
 	u.cancel()
 	return nil
 }
+
+func (u *UploadCompleter) emitSessionEndEvent(ctx context.Context, uploadData UploadMetadata) error {
+	var serverID, clusterName, user, login, hostname, namespace, serverAddr string
+	var interactive bool
+
+	// Get session events to find fields for constructed session end
+	sessionEvents, err := u.cfg.AuditLog.GetSessionEvents(defaults.Namespace, uploadData.SessionID, 0, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(sessionEvents) == 0 {
+		return nil
+	}
+	// Session start event is the first of session events
+	sessionStart := sessionEvents[0]
+
+	// Set variables
+	if sessionStart.GetType() == SessionStartEvent {
+		serverID = sessionStart.GetString(SessionServerHostname)
+		clusterName = sessionStart.GetString(SessionClusterName)
+		hostname = sessionStart.GetString(SessionServerHostname)
+		namespace = sessionStart.GetString(EventNamespace)
+		serverAddr = sessionStart.GetString(SessionServerAddr)
+		user = sessionStart.GetString(EventUser)
+		login = sessionStart.GetString(EventLogin)
+		if terminalSize := sessionStart.GetString(TerminalSize); terminalSize != "" {
+			interactive = true
+		}
+	} else {
+		return trace.BadParameter("invalid session, session start is not the first event")
+	}
```

Comment:
```suggestion
	if sessionStart.GetType() != SessionStartEvent {
		return trace.BadParameter("invalid session, session start is not the first event")
	}
	serverID = sessionStart.GetString(SessionServerHostname)
	clusterName = sessionStart.GetString(SessionClusterName)
	hostname = sessionStart.GetString(SessionServerHostname)
	namespace = sessionStart.GetString(EventNamespace)
	serverAddr = sessionStart.GetString(SessionServerAddr)
	user = sessionStart.GetString(EventUser)
	login = sessionStart.GetString(EventLogin)
	if terminalSize := sessionStart.GetString(TerminalSize); terminalSize != "" {
		interactive = true
	}
```

nit: avoid unnecessary nesting